prov/gni/include/gnix_tags.h

```diff
@@ -30,6 +30,95 @@
+ *
+ * A hash list implementation is not impossible, but can be computationally
+ * impractical. This is due to the way a search must occur if bits are ignored.
+ * The problem would be trivial if no bits were ignored, as the implementation
+ * could simply go to the bucket where the tag was hashed and search there.
+ * However, because ignored bits could hash to multiple, if not all, buckets,
+ * the tag matcher must search all buckets. It takes more time to compute all
+ * permutations of affected tags, than to just search each tag, so the default
+ * behavior is to search all buckets if a non-zero ignore field is provided.
+ *
+ * In the event that the ignore field is always zero, a hash list
+ * implementation is strictly better than a list implementation. However,
+ * once ignore bits are considered, the problem becomes more complex.
+ *
+ * Consider a hash list implementation where tags are split into K buckets,
+ * and tags are evenly distributed. If N tags are placed into the hash list in
+ * an evenly distributed manner, we can compute a best, average and worst
+ * case expectation for the hash list implementation.
+ *
+ * In the best case, the tag we are searching for is at the front of the list
+ * and only one tag in the system could match the search parameters, then
+ * we get the following best case analysis:
+ *
+ * When tags are evenly distributed, there should be ~(N/K) tags per bucket.
+ *
+ * Best(hash-list, middle list): ((n/k) * (k/ 2)) + 1 == (n/2) + 1
+ * Best(hash-list, first list): 1
+ *
+ *   We can assume that on average, we'll have to search half the buckets in
+ *   the best case, since the tags are evenly distributed. The very best case
+ *   is that the tag would be in the first bucket, giving us O(1) instead.
+ *   However, since tag location is determined on the hash, we have to assume
+ *   that we will have to search half the buckets first (N/K) * (K/2). Since
+ *   the element we are searching for is at the front of the list, it only
+ *   takes one search on that list.
+ *
+ * Average(hash-list): (N/K) * (K/2) + (N/(2K)) == (1/2) * (N + N/K)
+ *
+ *   Similar to the best case, we can assume we'll have to search through half
+ *   of the bucket when the ignore bits are set. Also, we can assume that the
+ *   entry we are searching for lies in the middle of one of the hash lists.
+ *   In this case, the average is not much worse than the best case.
+ *
+ * Worst(hash-list, last list): ((N/K) * K - 1) + (N/K) == N
+ *
+ *   The worst case is much more simple than the rest. It would be the last
+ *   element in the last list. This would be no worse than a linear search
+ *   through the list.
+ *
+ * All things considered, the hash list seems like it would be relatively
+ * reasonable with a more likely chance of finding entries faster than
+ * a standard linked list. However, a simple statistical analysis shows that
+ * that the best case is likely to be more expensive than the estimate.
```

Comment:
I removed the statistics after I realized they wouldn't be relevant.

Fair point.
-- Jim




On 1/27/16, 1:06 PM, "Sung-Eun Choi" <notifications@github.com> wrote:

>In 
>prov/gni/include/gnix_tags.h
><https://github.com/ofi-cray/libfabric-cray/pull/643#discussion_r51032661>
>:
>> + *
>> + *   Similar to the best case, we can assume we'll have to search
>>through half
>> + *   of the bucket when the ignore bits are set. Also, we can assume
>>that the
>> + *   entry we are searching for lies in the middle of one of the hash
>>lists.
>> + *   In this case, the average is not much worse than the best case.
>> + *
>> + * Worst(hash-list, last list): ((N/K) * K - 1) + (N/K) == N
>> + *
>> + *   The worst case is much more simple than the rest. It would be the
>>last
>> + *   element in the last list. This would be no worse than a linear
>>search
>> + *   through the list.
>> + *
>> + * All things considered, the hash list seems like it would be
>>relatively
>> + * reasonable with a more likely chance of finding entries faster than
>> + * a standard linked list. However, a simple statistical analysis
>>shows that
>> + * that the best case is likely to be more expensive than the estimate.
>Not to be too picky, but I don't think this is a statistical analysis.
>It's just an analysis.
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/643/files#r51032661>.
>
>